### ISSUE TITLE
Global Styles: Styles specificity and users expectations

### ISSUE THREAD

**Issue by Copons** [WordPress/gutenberg#44931]
CSS is hard; getting it right while translating it into a user-friendly interface is even harder.

In a GS world, the site's styles can originate from many, many sources: editor default, theme.json and style variations, Global Styles of various nature, individual block styles, custom CSS somewhere, etc.

It's a confusing scenario for the end user, even if tech-savvy, who might not know those sources, or understand how they interact with each other.

In a [recent test](https://github.com/Automattic/wp-calypso/issues/68916) we ran, we discovered that there's a subtle but important difference between "blocks" and "elements" that I personally never fully realized before.

While in most cases all our design tools affect blocks, **these items here affect actual HTML elements**:

<img width="277" alt="Screenshot 2022-10-13 at 11 43 50" src="https://user-images.githubusercontent.com/2070010/195577729-4aaf705e-b662-47a2-a817-84849dce6b62.png">

It is reasonable, technically sound, and translates visually what we used to do in code, in CSS or in the theme.json.

Though, when rendered, **element customizations have a lower priority than block customizations**.

Again, this is technically correct: HTML elements are more generic than blocks, and it's just right that block styles have precedence.

But it's also very confusing how we have element customization items in the main GS panel (so in a top-priority position) that, **when changed, might not do anything to the canvas**.

It is a common scenario that themes explicitly define styles for the Heading and Button blocks, which will always override any changes in the Global Styles' elements panels.

I wonder if we should **de-prioritize the location of the element panels**, since they are technically less _!important_ (CSS pun intended) than literally everything else in the GS panel.

---

cc a bunch of people related to GS that I know of: @carolinan @annezazu @scruffian @MaggieCabrera @mikachan @oandregal @jorgefilipecosta 

Context: this rabbit hole started [here](https://github.com/Automattic/wp-calypso/issues/68916) and continued [here](https://github.com/WordPress/twentytwentythree/issues/278), before finally landing into this issue.

**Comment by MaggieCabrera** [comment]
I'm not sure changing the order of the panels is a good solution for this. It will help the user that wants to edit a site, but maybe not one who is building one: a case where you want them editing from the global to the specific. I wonder if there's a way we can inform the user where changes from a block are coming from, so they don't need to think about the "cascade". Maybe this information could only be shown in Global Styles in the cases where overrides of styles are possible? I guess that can get convoluted really fast...

**Comment by MaggieCabrera** [comment]
I can also think of cases where elements are the only way (so far!) to edit the styles of a block partially: this is the case of the search and file blocks. Burying the option to change elements can end up being confusing or having the user change things twice, once in the block and then in the element, defeating the purpose of having the more global solution.

**Comment by mtias** [comment]
Yeah, the order and hierarchy is fine, it's just that we should display when there are Block Type styles set by the user or theme that are relevant for the category element you are in — if you are editing the "Headings" element we should show and allow access to the Heading, Post Title, Archive Title, Site Title blocks, etc.

Buttons is the same — when editing the generic element we should show the block type in context and as a shortcut.

**Comment by mtias** [comment]
Also related: #44361, #44420.

**Comment by Copons** [comment]
I know that notices are a lazy solution, but maybe reminding that element styles are overridden by block styles could be a good start. 🤔 

> it's just that we should display when there are Block Type styles set by the user or theme that are relevant for the category element you are in — if you are editing the "Headings" element we should show and allow access to the Heading, Post Title, Archive Title, Site Title blocks, etc.

I'm having a hard time visualizing this in my mind, especially in a way that wouldn't overload info.

**Comment by mtias** [comment]
Similar to the bottom part here:

<img width="442" alt="image" src="https://user-images.githubusercontent.com/548849/195662213-2538d447-fd01-47c2-a0cf-2c4331231af3.png">

**Comment by dsas** [comment]
https://github.com/WordPress/gutenberg/issues/45086 is related to this

**Comment by jasmussen** [comment]
Can we technically do something like what's suggested [here](https://github.com/WordPress/gutenberg/issues/44931#issuecomment-1277929989)? That is, show an ItemGroup where styles are "applied to"? 

It relates both to blocks and elements, but also a great deal to style inheritence.

Pinging some folks working on related bits: @MaggieCabrera @ramonjd @andrewserong

**Comment by ramonjd** [comment]
> Can we technically do something like what's suggested https://github.com/WordPress/gutenberg/issues/44931#issuecomment-1277929989? That is, show an ItemGroup where styles are "applied to"?

My thoughts also went to the global style inheritance work that @aaronrobertshaw was looking at.

As far as I know, it already does this but from the far end of Style Street™️ : it identifies the "winning" value from the inheritance chain (`theme > global style elements > global style blocks > block`). The "losing" values are forgotten about.

If I understand things correctly, here we want to know about some of the intermediary links in that chain, e.g., `global style elements > global style blocks` so we can tell, for example, on the headings screen if a font is being overridden by the global styles heading block. So we keep a record of the "losers" to compare.

The catch is, we only know which blocks an element affects for core blocks, and, I speculate, we'll need to maintain a  block + element relationship dictionary somewhere. [The inheritance experiment already does this to some extent](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/global-styles/inherited-value-context.js#L124). 

But outside that, nothing in a block tells us that `supports.color.link` says a block contains a link. An alternative is moving the declaration into block registration so that the block has this information itself, e.g., `"rendersElements": [ "heading" ]`. That's a pretty large job to weigh against the discoverability benefits, not to mention a separate project requiring some dedicated time/resources.

**Comment by aaronrobertshaw** [comment]
Thanks for the ping 👍 

I think Ramon captures most of my initial thoughts.

There is a dictionary of sorts on trunk already for the global styles inheritance display work. `getElementLayers` in the [inherited values context](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/global-styles/inherited-value-context.js#L124) maps a block to the element layers that paint it. At present, that is hand maintained and core blocks only. 

This issue seems to need the same map but backwards, elements to blocks. It should be pretty cheap to invert it.

An ItemGroup for core blocks would work but third party blocks rendering elements would not appear, as Ramon noted.  This is the same limit the inherited styles display already ships with.

I'd see this broken down into two pieces:

1. **Applies to list** - On an element screen, list the blocks that render the element, each linking to its Blocks screen.
2. **Overridden by per control** - We'd need the losing values Ramon spoke of. It might be possible to store the losing values that `resolveStyle` walks past so this could be an extension of that rather than a new system.

So, the step 1 above covers the support cases above and is worth doing on its own.  Step 2 could be done as a follow-up if Step 1 alone isn't enough.

**Comment by jasmussen** [comment]
Thanks both for chiming in. It is a bit inverting the chain, and it sounds like you both get it right. But to be sure myself, these two seem at odds:

> But outside that, nothing in a block tells us that supports.color.link says a block contains a link. 

> This issue seems to need the same map but backwards, elements to blocks. It should be pretty cheap to invert it.

The first line I read as, it's not possible. The latter suggests it is? I'm sure there's nuance, the only reason I ask is to note that **if there's no clean path forward, we can close this issue as not planned.**

However:

> Applies to list - On an element screen, list the blocks that render the element, each linking to its Blocks screen.

> So, the step 1 above covers the support cases above and is worth doing on its own.

This sounds good to me. Should we start here? Not urgent, and let's possibly change the title of this issue to be more representative if you agree.

**Comment by aaronrobertshaw** [comment]
I see it a little differently and the two comments aren't at odds.

> But outside that, nothing in a block tells us that supports.color.link says a block contains a link.

My read is this is aimed at being able to automatically determine which blocks render elements where the block styles would override the element styles.

> This issue seems to need the same map but backwards, elements to blocks. It should be pretty cheap to invert it.

This was referencing the arbitrary map that is already in place for the display of inherited global styles on blocks. The need for the arbitrary map in effect reinforces the first comment i.e. we haven't found a clean way from block type registration to determine the relationship between blocks and elements.

> This sounds good to me. Should we start here? Not urgent, and let's possibly change the title of this issue to be more representative if you agree.

I'm not against making a start here. It might be good to confirm what the design should be. The [earlier design](https://github.com/WordPress/gutenberg/issues/44931#issuecomment-1277929989) does seem to indicate that what is applied to a block (Step 2) is rendered as a swatch on the block. It might be possible but a small MVP might be to just omit the swatch, would that be an acceptable stepping stone?

**Comment by jasmussen** [comment]
> It might be possible but a small MVP might be to just omit the swatch, would that be an acceptable stepping stone?

Yes, most definitely and absolutely, especially if that's simplish to make. If the cost of seeing the swatch is a single click, the swatch may not be needed at all. (Assuming I understood you correctly.)

**Comment by ramonjd** [comment]
> My read is this is aimed at being able to automatically determine which blocks render elements where the block styles would override the element styles.

Thank you for translating my babble. This is what I was going for 😄 pursuant to my rando block registration idea: 

> block registration so that the block has this information [which elements it renders] itself, e.g., "rendersElements": [ "heading" ].

**Comment by aaronrobertshaw** [comment]
Sounds like we have a plan 👍 

There's some more to do on the display of inherited global styles so while I'm in the area I'll take a run at this as well.

**Comment by jasmussen** [comment]
Noting that in testing another issue, there's [a little related information](https://github.com/WordPress/gutenberg/issues/54520#issuecomment-5791657476) in it worth knowing for any work here.
